### PR TITLE
Feature: Fann pet leveling costs Coins/XP and XP/Bits

### PR DESCRIPTION
**Pull request by minhperry**
## What
https://discord.com/channels/997079228510117908/1192656223741427804

<details>
<summary>Images</summary>

image

image

</details>

## Changelog New Features
+ Added Fann leveling cost coins per XP and XP per bit. - minhperry

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/config/features/inventory/FannCostConfig.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+import at.hannibal2.skyhanni.config.FeatureToggle;
+import com.google.gson.annotations.Expose;
+import io.github.notenoughupdates.moulconfig.annotations.ConfigEditorBoolean;
+import io.github.notenoughupdates.moulconfig.annotations.ConfigOption;
+
+public class FannCostConfig {
+    
+    @Expose
+    @ConfigOption(name = "Coins/XP", desc = "Shows coins per XP in Fann menu")
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/inventory/FannCostConfig.java]:
```suggestion
    @ConfigOption(name = "Coins/XP", desc = "Shows coins per XP in Fann menu.")
```

File: src/main/java/at/hannibal2/skyhanni/config/features/inventory/FannCostConfig.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    @ConfigEditorBoolean
+    @FeatureToggle
+    public boolean coinsPerXP = false;
+
+    @Expose
+    @ConfigOption(name = "XP/Bits", desc = "Shows XP per bit in Fann menu")
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/inventory/FannCostConfig.java]:
```suggestion
    @ConfigOption(name = "XP/Bits", desc = "Shows XP per bit in Fann menu.")
```

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        for (line in lore) {
+            val linePlain = line.removeColor()
+            val matcher = matcher(linePlain)
+            if (matcher.find()) {
+                val res = matcher.group(1)
+                return func(res)
+            }
+        }
+        return null
+    }
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt]: im sure a function that does this is already in RegexUtils

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        return this.group(key)?.toInt() ?: 0
+    }
+
+    private fun String._toDouble(): Double {
+        return this.replace(",", "").toDouble()
+    }
+
+    private enum class TrainingMode {
+        DAY_COUNT,
+        UNTIL_LEVEL
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt]:
```suggestion
        UNTIL_LEVEL,
```

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        return null  // Return null if no valid duration is found
+    }
+
+    private fun Matcher.getIntOrZero(key: String): Int {
+        return this.group(key)?.toInt() ?: 0
+    }
+
+    private fun String._toDouble(): Double {
+        return this.replace(",", "").toDouble()
+    }
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt]: There should be stuff that does this in NumberUtil

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+                // Return the total duration as a Double representing days
+                return totalDays
+            }
+        }
+        return null  // Return null if no valid duration is found
+    }
+
+    private fun Matcher.getIntOrZero(key: String): Int {
+        return this.group(key)?.toInt() ?: 0
+    }
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt]: RegexUtils has groupOrNull which then you can elvis operator to 0

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    private fun List<String>.getTrainingType(): TrainingType? {
+        return trainingTypePattern.read(this) { typestr ->
+            TrainingType.entries.firstOrNull { it.type == typestr }
+        }
+    }
+
+    private fun List<String>.getDuration(): Double? {
+        for (line in this) {
+            val linePlain = line.removeColor()
+            val matcher = durationPattern.matcher(linePlain)
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt]: should use Pattern.firstMatcher(list)

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+                tooltip.insertLineAfter(bitsPattern, "   §6XP/Bit: ${xpPerBit.roundTo(2)}")
+
+            }
+        }
+    }
+
+    @SubscribeEvent
+    fun onInventoryUpdate(event: InventoryUpdatedEvent) {
+        if (!trainingSlotInventoryPattern.matches(event.inventoryName.removeColor())) return
+        val slot24 = event.inventoryItems[24]
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt]: instead of assigning a variable then checking if it is null, just check if slot 24 is null wiht an elvis operator to then return from the function

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+    private fun MutableList<String>.insertLineAfter(pattern: Pattern, content: String) {
+        val iter = this.listIterator()
+        while (iter.hasNext()) {
+            val line = iter.next().removeColor()
+            if (pattern.matcher(line).find()) {
+                iter.add(content)
+            }
+        }
+    }
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt]: This seems like a cool function that could be used elsewhere in the mod, maybe move it to a utils class such as collection utils

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+        "Desired Level: (200|1?[0-9]?[0-9])",
+    )
+    private val userInputPattern by patternGroup.pattern(
+        "slot24.name.input",
+        "User Input",
+    )
+    private val trainingTypePattern by patternGroup.pattern(
+        "training.type",
+        "Type: (Free|Light|Moderate|Expert|Ultra|Turbo!)",
+    )
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt]: all of these need regex tests, they also need named pattern groups and none of them should include triple quotes

**Review by CalMWolfs — CHANGES_REQUESTED** [review]

**Comment by minhperry** [comment]
I don't think I can continue this any time soon, feel free to continue

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+    /**
+     * REGEX-TEST: Will take: 1d 0h 0m 0s
+     * REGEX-TEST: Will take: 3d 11h 10m 35s
+     * REGEX-TEST: Will take: 0d 0h 0m 1s
+     * REGEX-TEST: Will take: 493d 19h 49m 59s
+     */
+    private val durationPattern by patternGroup.pattern(
+        "training.duration.pattern",
+        "Will take: (?<day>\\d+)d (?<hr>\\d{1,2})h (?<min>\\d{1,2})m (?<sec>\\d{1,2})s",
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt]: We do have time formatting functions already. Please just check against `Will Take: (?<time>.*)`

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+        while (iter.hasNext()) {
+            val line = iter.next().removeColor()
+            if (pattern.matcher(line).find()) {
+                iter.add(content)
+            }
+        }
+    }
+
+    // To avoid duplicate code
+    private fun <T> Pattern.read(lore: List<String>, func: (String) -> T): T? {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt]: Please do not work with the group number of patterns, rather use the group name instead.

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+            TrainingType.entries.firstOrNull { it.type == typestr }
+        }
+    }
+
+    private fun List<String>.getDuration(): Double? {
+        for (line in this) {
+            val linePlain = line.removeColor()
+            val matcher = durationPattern.matcher(linePlain)
+            if (matcher.find()) {
+                val days = matcher.groupOrNull("day")?.toInt() ?: 0
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt]: can all be replaced with `TimeUtils.getDuration()`

**Review by hannibal002 — CHANGES_REQUESTED** [review]

**Comment by minhperry** [comment]
what...

File: src/main/java/at/hannibal2/skyhanni/utils/CollectionUtils.kt [hunk 14]
```diff
@@ -489,4 +491,20 @@ object CollectionUtils {
...
         if (index < size) add(index, element) else add(element)
     }
+
+    /**
+     * Insert content after a line that matches the given pattern.
+     *
+     * @param pattern the pattern to match
+     * @param content the content to insert
+     */
+    fun MutableList<T>.insertLineAfter(pattern: Pattern, content: T) {
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/utils/CollectionUtils.kt]:
```suggestion
    fun <T> MutableList<T>.insertLineAfter(pattern: Pattern, content: T) {
```

**Comment by minhperry** [comment]
this was not tested with bits yet because I don't have the Turbo option unlocked.

**Comment by minhperry** [comment]
Finally coming back to this reviving it from the graveyard

**Comment by hannibal002** [comment]
Welcome back o/

**Comment by minhperry** [comment]
Closing this because #3969 use a cached approach which is less costly
